### PR TITLE
core/txpool/legacypool: remove a redundant heap.Init

### PR DESCRIPTION
**Pull request by colinlyguo**
Found a redundant `heap.Init(m.index)` when learning the implementation of tx pool. this would save `O(threshold)` in each `Cap()` call.

Because `*m.index` is sorted increasingly and truncated the largest `size - threshold` values (same as `m.Pop()` for `size - threshold` times), or adding `heap.Init(m.index)` for completeness sake?

### CODE REVIEW

**Comment by fjl** [comment]
This looks incorrect to me, but I'm slightly unsure. `heap.Init` restores the 'heap property' which is:

```go
!h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len()
```

The actual storage of a heap slice is NOT simply a sorted slice, it's a binary tree. So I think we can't just use the sorted slice as the heap.

**Comment by colinlyguo** [comment]
> This looks incorrect to me, but I'm slightly unsure. `heap.Init` restores the 'heap property' (index starting from `0`) which is:
> 
> ```go
> !h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len()
> ```
> 
> The actual storage of a heap slice is NOT simply a sorted slice, it's a binary tree. So I think we can't just use the sorted slice as the heap.

Thanks for the discussion. I will demonstrate that a sorted slice satisfies the invariance of a min-heap, which can be expressed as:
```
!h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len()
```

This invariant means that for any valid index `i` within the heap, the value at index `i` should not be greater than the values at its children's indices `2*i+1` and `2*i+2`, provided that these children indices are within the bounds of the heap.

In a stored slice:
```
for any j >= i, a[j] >= a[i]
```
This directly implies the heap invariance because if you pick any index `i` and its corresponding children indices `2*i+1` and `2*i+2`, you can be assured that `a[i]` will be less than or equal to `a[2*i+1]` and `a[2*i+2]`, as `2*i+1` and `2*i+2` are both greater than `i` and, thus, their values are greater than or equal to `a[i]` in a sorted array.

**Comment by holiman** [comment]
I agree, have seen this before. A sorted slice is a subset of a heap. A heap is just slightly more free, whereas the sorted slice is strict.

**Comment by holiman** [comment]
I agree that the change seems correct, but otoh:
- Not sure any testcase covers it (i tested intentionally reverse-sorting the `m.index` and running the tests in `list_test.go`)
- The change makes the code less robust by depending on an implicit previous event  -- more sensitive to refactorings.
- Not sure how large (if any) the processing gain is

So I'm kind of leaning towards letting it be as it is... (?)

**Comment by colinlyguo** [comment]
> I agree that the change seems correct, but otoh:
> 
> * Not sure any testcase covers it (i tested intentionally reverse-sorting the `m.index` and running the tests in `list_test.go`)
> * The change makes the code less robust by depending on an implicit previous event  -- more sensitive to refactorings.
> * Not sure how large (if any) the processing gain is
> 
> So I'm kind of leaning towards letting it be as it is... (?)

Agree with point 2, adding a `heap.Init` is helpful for logic completeness.

btw, I added a benchmark test, removing `heap.Init` can decrease `4.27%` op time when popping out a transaction in a heap with 32 transactions (under an account).
```
colin@192 legacypool % go test -v -bench=BenchmarkListCapOneTx -run=^$ 
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/core/txpool/legacypool
BenchmarkListCapOneTx
BenchmarkListCapOneTx-12          597567              2013 ns/op
PASS
ok      github.com/ethereum/go-ethereum/core/txpool/legacypool  38.480s
colin@192 legacypool % go test -v -bench=BenchmarkListCapOneTx -run=^$
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/core/txpool/legacypool
BenchmarkListCapOneTx
BenchmarkListCapOneTx-12          642597              1927 ns/op
PASS
ok      github.com/ethereum/go-ethereum/core/txpool/legacypool  41.322s
```

**Comment by fjl** [comment]
I think it will have a larger effect if the transactions list is much longer, as it usually is for real world use cases. `heap.Init` has a complexity of `O(n)`.

**Comment by colinlyguo** [comment]
> I think it will have a larger effect if the transactions list is much longer, as it usually is for real world use cases. `heap.Init` has a complexity of `O(n)`.

Make sense, some logs about the exact time inside Cap():
```
total number 100000 threshold 99999
Check limit duration: 125ns
Sort duration: 10.84525ms
Drop duration: 4.875µs
Heap init duration: 239.75µs
Cache duration: 250ns
Total Cap function duration: 11.106458ms
```
The `sort.Sort(*m.index)` contributes the most latency. and `heap.Init` contributes the second.

**Comment by colinlyguo** [comment]
> I think it will have a larger effect if the transactions list is much longer, as it usually is for real world use cases. `heap.Init` has a complexity of `O(n)`.

refactored this function in [e8b23da](https://github.com/ethereum/go-ethereum/pull/28910/commits/e8b23da4cd6c1e4e8fd6fdb239bc486ba2debf5b)
It's faster than original `2013 ns/op`, only consuming `591.7 ns/op` now.
The theoretical complexity is also better than the previous implementation.
```
colin@192 legacypool % go test -v -bench=BenchmarkListCapOneTx -run=^$
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/core/txpool/legacypool
BenchmarkListCapOneTx
BenchmarkListCapOneTx-12         2066755               591.7 ns/op
PASS
ok      github.com/ethereum/go-ethereum/core/txpool/legacypool  189.050s
```

**Comment by holiman** [comment]
So if I read it correctly, sorting is `10.8ms`, and a heap-init of the already sorted list is `0.2ms` ?

File: core/txpool/legacypool/list.go [hunk 1]
```diff
@@ -160,13 +160,11 @@ func (m *sortedMap) Cap(threshold int) types.Transactions {
 	// Otherwise gather and drop the highest nonce'd transactions
 	var drops types.Transactions
 
-	sort.Sort(*m.index)
 	for size := len(m.items); size > threshold; size-- {
-		drops = append(drops, m.items[(*m.index)[size-1]])
-		delete(m.items, (*m.index)[size-1])
+		dropIdx := m.index.Pop().(uint64)
```
Comment (fjl) [inline core/txpool/legacypool/list.go:164]: The documentation of container/heap says: 

> Note that [Push](https://pkg.go.dev/container/heap#Push) and [Pop](https://pkg.go.dev/container/heap#Pop) in this interface are for package heap's implementation to call. To add and remove things from the heap, use [heap.Push](https://pkg.go.dev/container/heap#Push) and [heap.Pop](https://pkg.go.dev/container/heap#Pop).

So you need to call `heap.Pop(m.index)` here.

Comment (holiman) [inline core/txpool/legacypool/list.go:164]: Doens't this negate the previous algo? popping removes from the front of the list, not the back??

Comment (colinlyguo) [reply core/txpool/legacypool/list.go:164]: Yeah. forgive me. this is a mistake. reverted.

Comment (colinlyguo) [reply core/txpool/legacypool/list.go]: the sort is unavoidable here.

File: core/txpool/legacypool/list.go [hunk 2]
```diff
@@ -166,7 +166,6 @@ func (m *sortedMap) Cap(threshold int) types.Transactions {
 		delete(m.items, (*m.index)[size-1])
 	}
 	*m.index = (*m.index)[:threshold]
-	heap.Init(m.index)
```
Comment (karalabe) [inline core/txpool/legacypool/list.go:169]: Could you add a comment after `sort.Sort(*m.index)` along the lines of `// the sorted array is a heap, so no need to reheap after deleting tail items`

**Comment by colinlyguo** [comment]
> So if I read it correctly, sorting is `10.8ms`, and a heap-init of the already sorted list is `0.2ms` ?

Yeah, I think it's because the `heap.Init()` would just iterate all non-leaf heap nodes without changing the heap structure.

**Review by holiman — APPROVED** [review]
LGTM

File: core/txpool/legacypool/list.go [hunk 3]
```diff
@@ -160,13 +160,12 @@ func (m *sortedMap) Cap(threshold int) types.Transactions {
 	// Otherwise gather and drop the highest nonce'd transactions
 	var drops types.Transactions
 
-	sort.Sort(*m.index)
+	sort.Sort(*m.index) // the sorted array is a heap, so no need to reheap after deleting tail items
```
Comment (holiman) [inline core/txpool/legacypool/list.go:163]: I suspect an even bigger gain would be to replace this with 
```suggestion
	slices.Sort(*m.index) // the sorted array is a heap, so no need to reheap after deleting tail items
```

**Comment by colinlyguo** [comment]
Twice faster after changing the sorting function. quite thanks to @holiman 
```
colin@192 legacypool % go test -v -bench=BenchmarkListCapOneTx -run=^$
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/core/txpool/legacypool
BenchmarkListCapOneTx
BenchmarkListCapOneTx-12         1214308               975.7 ns/op
PASS
ok      github.com/ethereum/go-ethereum/core/txpool/legacypool  121.775s
```

**Review by fjl — APPROVED** [review]

**Comment by fjl** [comment]
I'm merging this because we're now running into the test timeout that has been plagueing the CI builder recently. The change itself is OK now.

**Pull request merged**